packages/web-components/src/donut-chart/donut-chart.ts

```diff
@@ -0,0 +1,344 @@
+          }
+        }
+        for (let j = 0; j < buttons.length; j++) {
+          const legendRect = (buttons[j].getElementsByClassName('legendRect') as HTMLCollectionOf<HTMLDivElement>)[0];
+          const legendText = (buttons[j].getElementsByClassName('legendText') as HTMLCollectionOf<HTMLDivElement>)[0];
+
+          if (j !== i) {
+            legendRect.style['backgroundColor'] = 'transparent';
+            legendText.style['opacity'] = '0.67';
+          } else {
+            legendRect.style['backgroundColor'] = legends[j].color;
+            legendText.style['opacity'] = '1';
+          }
+        }
+      });
+      buttons[i].addEventListener('blur', () => {
+        if (this._selectedLegend !== '') {
+          return;
+        }
+
+        for (let j = 0; j < arcs.length; j++) {
+          arcs[j].style['opacity'] = '1';
+        }
+        for (let j = 0; j < buttons.length; j++) {
+          const legendRect = (buttons[j].getElementsByClassName('legendRect') as HTMLCollectionOf<HTMLDivElement>)[0];
+          legendRect.style['backgroundColor'] = legends[j].color;
+
+          const legendText = (buttons[j].getElementsByClassName('legendText') as HTMLCollectionOf<HTMLDivElement>)[0];
+          legendText.style['opacity'] = '1';
+        }
+      });
+      buttons[i].addEventListener('click', () => {
+        if (this._selectedLegend === legends[i].title) {
+          this._selectedLegend = '';
+
+          for (let j = 0; j < arcs.length; j++) {
+            arcs[j].style['opacity'] = '1';
+          }
+          for (let j = 0; j < buttons.length; j++) {
+            const legendRect = (buttons[j].getElementsByClassName('legendRect') as HTMLCollectionOf<HTMLDivElement>)[0];
+            legendRect.style['backgroundColor'] = legends[j].color;
+
+            const legendText = (buttons[j].getElementsByClassName('legendText') as HTMLCollectionOf<HTMLDivElement>)[0];
+            legendText.style['opacity'] = '1';
+          }
+        } else {
+          this._selectedLegend = legends[i].title;
+
+          for (let j = 0; j < arcs.length; j++) {
+            if (arcs[j].getAttribute('data-id') !== buttons[i].textContent) {
+              arcs[j].style['opacity'] = '0.1';
+            } else {
+              arcs[j].style['opacity'] = '1';
+            }
+          }
+          for (let j = 0; j < buttons.length; j++) {
+            const legendRect = (buttons[j].getElementsByClassName('legendRect') as HTMLCollectionOf<HTMLDivElement>)[0];
+            const legendText = (buttons[j].getElementsByClassName('legendText') as HTMLCollectionOf<HTMLDivElement>)[0];
+
+            if (j !== i) {
+              legendRect.style['backgroundColor'] = 'transparent';
+              legendText.style['opacity'] = '0.67';
+            } else {
+              legendRect.style['backgroundColor'] = legends[j].color;
+              legendText.style['opacity'] = '1';
+            }
+          }
+        }
```

Comment:
Same suggestion as in HorizontalBarChart, there's a lot of repetition here and it looks like it could be pulled out into a single functional call.

Reply:
fixed in #33149